File: bind/frame64_example.cc
// frame gen/sync 64 C++ example
#include <stdio.h>
#include "fg64.hh"
#include "fs64.hh"

namespace dsp = liquid;

// context for passing to callback
struct context_s { int num_detected; };

// user-defined callback
int my_callback(
        unsigned char *  _header,
        int              _header_valid,
        unsigned char *  _payload,
        unsigned int     _payload_len,
        int              _payload_valid,
        framesyncstats_s _stats,
        void *           _userdata)
{
    printf("callback invoked! payload: %s\n", _payload_valid ? "pass" : "FAIL");
    ((context_s *)_userdata)->num_detected++;
    return 0;
}

int main()
{
    // generate header and payload bytes
    uint8_t header[8], payload[64];
    for (auto i=0; i< 8; i++) header[i]  = rand() & 0xff;
    for (auto i=0; i<64; i++) payload[i] = rand() & 0xff;

    // create frame generator and write frame to buffer
    dsp::fg64 fg;
    unsigned int frame_len = fg.get_frame_length();
    std::complex<float> frame_buf[frame_len];
    fg.execute(header, payload, frame_buf);

    // create frame synchronizer and receive frame
    context_s context = {.num_detected=0};
    dsp::fs64 fs(my_callback, &context);
    fs.execute(frame_buf, frame_len);
    printf("number of packets detected: %u\n", context.num_detected);
    return 0;
}

File: bind/frame64_example.py
#!/usr/bin/env python3
'''demonstrate framesync64 detection using class with callback function'''
import argparse
import liquid as dsp, numpy as np, matplotlib.pyplot as plt
p = argparse.ArgumentParser(description=__doc__)
p.add_argument('-nodisplay', action='store_true', help='disable display')
args = p.parse_args()

# generate a frame
fg      = dsp.fg64()
header  = np.random.randint(256, size=fg.header_len,  dtype=np.uint8)
payload = np.random.randint(256, size=fg.payload_len, dtype=np.uint8)
#payload = np.arange(16, dtype=np.uint8)
#payload = 'hello, world!'
frame   = fg.execute(header, payload)

def callback(context,header,payload,stats):
    print('frame detected!')
    print(payload)
    #print(''.join([chr(c) for c in payload]))
    context.update(stats)

# create frame synchronizer and receive frame
context = {'syms':np.zeros((0,))}
fs = dsp.fs64(callback,context)
fs.execute(frame)

# compute spectral response
nfft = 2400
psd = 20*np.log10(np.abs(np.fft.fftshift(np.fft.fft(frame, nfft))))

# plot frame and its spectrum
fig, ax = plt.subplots(2,figsize=(8,8))
t = np.arange(fg.frame_len)
ax[0].plot(t, np.real(frame), t, np.imag(frame))
ax[0].set_xlabel('Delay [samples]')
ax[0].set_ylabel('Frame Samples')
ax[0].grid(True, zorder=5)
ax[1].plot(np.arange(nfft)/nfft-0.5, psd)
ax[1].set_xlabel('Normalized Frequency [f/F_s]')
ax[1].set_ylabel('Power Spectral Density [dB]')
ax[1].grid(True, zorder=5)

fig, ax = plt.subplots(1,figsize=(8,8))
ax.plot(np.real(context['syms']),np.imag(context['syms']),'x')
ax.set_xlabel('real')
ax.set_ylabel('imag')
plt.xlim((-1.2,1.2))
plt.ylim((-1.2,1.2))
ax.grid(True)

if not args.nodisplay:
    plt.show()

